### ISSUE TITLE
Supporting CASRenew when <Directory> and .htaccess directives cannot be used?

### ISSUE THREAD

**Issue by bkosborne** [apereo/mod_auth_cas#195]
We have Apache configured a proxy for some backend services, similar to this basic setup:

```
<VirtualHost samplesite.com:443>
ServerName samplesite.com
DocumentRoot /var/www

...

CASCookiePath /var/cache/apache2/mod_auth_cas/
CASLoginURL https://cas.example.com/cas/login

<Location />
        Authtype CAS
        require valid-user
        CASAuthNHeader Cas-User
</Location>

ProxyPreserveHost On
ProxyPass        / http://localhost:8000/
ProxyPassReverse / http://localhost:8000/
</VirtualHost>
```

The CASRenew directive is only valid in <Directory> or .htaccess, but neither of those can be used in our situation where the Apache server is just acting as a reverse proxy for the protected application.

I thought maybe we could just append ?renew=true to the end of the CASLoginURL, but looks like that won't work because  [the C function that constructs the redirect to the CAS server](https://github.com/apereo/mod_auth_cas/blob/master/src/mod_auth_cas.c#L643) assumes that the login URL doesn't have any query params in it already.

I suppose this is really a feature request to allow CASRenew to be set at the Server/VirtualHost level as well?

Thanks

**Comment by dhawes** [comment]
Try adding "CASRenew /" to your Location block. I think that will get you what you want.

**Comment by dhawes** [comment]
I have a similar proxy config that works.

CASRenew is defined as ACCESS_CONF|OR_AUTHCFG [1], meaning:

ACCESS_CONF: Allow in .conf files (not .htaccess) inside <Directory> or <Location>
OR_AUTHCFG: Allow in .conf files and .htaccess when AllowOverride AuthConfig is set

What do your debug logs say (CASDebug On)?

[1] https://github.com/apereo/mod_auth_cas/blob/master/src/mod_auth_cas.c#L2892

**Comment by bkosborne** [comment]
@dhawes  Here's the output of the debug log the an authentication attempt. I obscured some hostnames/IP addresses. We are using SAML auth option, and it looks like there's possibly a bug in that code. When the ticket was attempting to be validated, it looked like the "renew" param get passed along to the end of the ticket?

```
[Thu Oct 01 15:22:55.578606 2020] [socache_shmcb:debug] [pid 113663] mod_socache_shmcb.c(485): AH00831: socache_shmcb_store (0x4b -> subcache 11)
[Thu Oct 01 15:22:55.578696 2020] [socache_shmcb:debug] [pid 113663] mod_socache_shmcb.c(810): AH00847: insert happened at idx=0, data=(0:32)
[Thu Oct 01 15:22:55.578706 2020] [socache_shmcb:debug] [pid 113663] mod_socache_shmcb.c(815): AH00848: finished insert, subcache: idx_pos/idx_used=0/1, data_pos/data_used=0/214
[Thu Oct 01 15:22:55.578712 2020] [socache_shmcb:debug] [pid 113663] mod_socache_shmcb.c(506): AH00834: leaving socache_shmcb_store successfully
[Thu Oct 01 15:22:55.580548 2020] [ssl:debug] [pid 113663] ssl_engine_kernel.c(225): [client 10.x.x.x:58284] AH02034: Initial (No.1) HTTPS request received for child 2 (server myapp.sample.com:443)
[Thu Oct 01 15:22:55.580660 2020] [authz_core:debug] [pid 113663] mod_authz_core.c(809): [client 10.x.x.x:58284] AH01626: authorization result of Require valid-user : denied (no authenticated user yet)
[Thu Oct 01 15:22:55.580669 2020] [authz_core:debug] [pid 113663] mod_authz_core.c(809): [client 10.x.x.x:58284] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet)
[Thu Oct 01 15:22:55.580685 2020] [:debug] [pid 113663] mod_auth_cas.c(2058): [client 10.x.x.x:58284] Entering cas_authenticate()
[Thu Oct 01 15:22:55.580723 2020] [:debug] [pid 113663] mod_auth_cas.c(580): [client 10.x.x.x:58284] CAS Service 'https%3a%2f%2fmyapp.sample.com%2f'
[Thu Oct 01 15:22:55.580728 2020] [:debug] [pid 113663] mod_auth_cas.c(528): [client 10.x.x.x:58284] entering getCASLoginURL()
[Thu Oct 01 15:22:55.580735 2020] [:debug] [pid 113663] mod_auth_cas.c(505): [client 10.x.x.x:58284] entering getCASGateway()
[Thu Oct 01 15:22:55.580739 2020] [:debug] [pid 113663] mod_auth_cas.c(595): [client 10.x.x.x:58284] entering redirectRequest()
[Thu Oct 01 15:22:55.580744 2020] [:debug] [pid 113663] mod_auth_cas.c(607): [client 10.x.x.x:58284] Adding outgoing header: Location: https://casserver.sample.com/cas/login?service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597678 2020] [ssl:debug] [pid 113662] ssl_engine_kernel.c(225): [client 172.x.x.x:60687] AH02034: Subsequent (No.2) HTTPS request received for child 1 (server myapp.sample.com:443), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597751 2020] [authz_core:debug] [pid 113662] mod_authz_core.c(809): [client 172.x.x.x:60687] AH01626: authorization result of Require valid-user : denied (no authenticated user yet), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597759 2020] [authz_core:debug] [pid 113662] mod_authz_core.c(809): [client 172.x.x.x:60687] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597774 2020] [:debug] [pid 113662] mod_auth_cas.c(2058): [client 172.x.x.x:60687] Entering cas_authenticate(), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597787 2020] [:debug] [pid 113662] mod_auth_cas.c(652): [client 172.x.x.x:60687] Modified r->args (now ''), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.597850 2020] [:debug] [pid 113662] mod_auth_cas.c(1761): [client 172.x.x.x:60687] entering getResponseFromServer(), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.598010 2020] [:debug] [pid 113662] mod_auth_cas.c(580): [client 172.x.x.x:60687] CAS Service 'https%3a%2f%2fmyapp.sample.com%2f', referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.817800 2020] [:debug] [pid 113662] mod_auth_cas.c(1838): [client 172.x.x.x:60687] Validation response: <?xml version="1.0" encoding="UTF-8"?><SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Body><saml1p:Response xmlns:saml1p="urn:oasis:names:tc:SAML:1.0:protocol" IssueInstant="2020-10-01T19:22:55.811Z" MajorVersion="1" MinorVersion="1" Recipient="https://myapp.sample.com/" ResponseID="_db8b1092f6dd575b52ad63a97d12ce04"><saml1p:Status><saml1p:StatusCode Value="saml1p:RequestDenied"/><saml1p:StatusMessage>ticket 'ST-447552-ghRs43tboZbFsghD2eqm-auth-a&amp;renew=true' not recognized</saml1p:StatusMessage></saml1p:Status></saml1p:Response></SOAP-ENV:Body></SOAP-ENV:Envelope>, referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.818599 2020] [:debug] [pid 113662] mod_auth_cas.c(1436): [client 172.x.x.x:60687] entering isValidCASTicket(), referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.818611 2020] [:debug] [pid 113662] mod_auth_cas.c(1442): [client 172.x.x.x:60687] MOD_AUTH_CAS: response = <?xml version="1.0" encoding="UTF-8"?><SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Body><saml1p:Response xmlns:saml1p="urn:oasis:names:tc:SAML:1.0:protocol" IssueInstant="2020-10-01T19:22:55.811Z" MajorVersion="1" MinorVersion="1" Recipient="https://myapp.sample.com/" ResponseID="_db8b1092f6dd575b52ad63a97d12ce04"><saml1p:Status><saml1p:StatusCode Value="saml1p:RequestDenied"/><saml1p:StatusMessage>ticket 'ST-447552-ghRs43tboZbFsghD2eqm-auth-a&amp;renew=true' not recognized</saml1p:StatusMessage></saml1p:Status></saml1p:Response></SOAP-ENV:Body></SOAP-ENV:Envelope>, referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
[Thu Oct 01 15:22:55.818830 2020] [:error] [pid 113662] [client 172.x.x.x:60687] MOD_AUTH_CAS: unsupported SAML StatusCode, referer: https://casserver.sample.com/cas/login?locale=en&service=https%3a%2f%2fmyapp.sample.com%2f&renew=true
```

**Comment by bkosborne** [comment]
And here's the actual config:

```
<VirtualHost *:443>
  ServerName myapp.sample.edu
  DocumentRoot "/var/www/ssl"
  AllowEncodedSlashes nodecode

  <Directory "/var/www/ssl">
    AllowOverride None
    Require all granted
  </Directory>

  <Location "/">
    Require valid-user
    AuthType CAS
    RequestHeader set REMOTE_USER %{REMOTE_USER}s
    CASAuthNHeader On
  </Location>

  RequestHeader set X-Forwarded-Proto "https"
  RequestHeader set X-Forwarded-Port "443"

  ProxyRequests Off
  ProxyPreserveHost On
  ProxyAddHeaders On
  ProxyPass // http://backend.sample.com:10011/ connectiontimeout=300 timeout=300 nocanon
  ProxyPassReverse // http://backend.sample.com:10011/

  CASIdleTimeout 300
  CASLoginURL https://casserver.sample.com/cas/login
  CASValidateURL https://casserver.sample.com/cas/samlValidate
  CASValidateSAML On
  CASAttributePrefix X-CAS-
</VirtualHost>
```

**Comment by bkosborne** [comment]
Oof... looking thru the issue queue, I found this old PR that fixes the problem: https://github.com/apereo/mod_auth_cas/pull/138

It was apparently released as part of 1.2, and we're using 1.1! I think that's probably our issue!
